### PR TITLE
[close #265] Add test case for tls

### LINKED ISSUE #265: cdc: tls needs test case
## Feature Request

**Is your feature request related to a problem? Please describe:**

TiKV-CDC TLS needs test case, include server tls test & cli tls test.

**Describe the feature you'd like:**

TiKV-CDC TLS needs test case, include server tls test & cli tls test.

**Describe alternatives you've considered:**

**Teachability, Documentation, Adoption, Migration Strategy:**

### PR DESCRIPTION
**Pull request by zeminzhou**
### What problem does this PR solve?
Add test case for TLS

Issue Number: close #265

Problem Description: **TBD**
TiKV-CDC TLS needs test case, include server tls test & cli tls test.

### What is changed and how does it work?
1. Add test case for http api with tls and fix related bugs
2. Add test case for cli with tls and fix related bugs

### Code changes

- Has exported function/method change
- Has exported variable/fields change

### Check List for Tests

This PR has been tested by at least one of the following methods:
- Unit test
- Integration test
- Manual test (add detailed scripts or steps below)

### CODE REVIEW

**Comment by zeminzhou** [comment]
PTAL~

**Review by haojinming — APPROVED** [review]
LGTM

File: cdc/tests/integration_tests/cli_tls/run.sh [hunk 1]
```diff
@@ -0,0 +1,115 @@
+#!/bin/bash
+
+set -eu
```
Comment (pingyu) [inline cdc/tests/integration_tests/cli_tls/run.sh:3]: Suggest to add `-o pipefail`

File: cdc/tests/integration_tests/cli_tls/run.sh [hunk 2]
```diff
@@ -0,0 +1,115 @@
...
+
+CUR=$(cd "$(dirname "${BASH_SOURCE[0]}")" && pwd)
+source $CUR/../_utils/test_prepare
+WORK_DIR=$OUT_DIR/$TEST_NAME
+CDC_BINARY=tikv-cdc.test
+SINK_TYPE=$1
+TLS_DIR=$(cd $CUR/../_certificates && pwd)
+UP_PD=https://$TLS_PD_HOST:$TLS_PD_PORT
+DOWN_PD=http://$DOWN_PD_HOST:$DOWN_PD_PORT
```
Comment (pingyu) [inline cdc/tests/integration_tests/cli_tls/run.sh:12]: Suggest to test on downstream with TLS. It's very likely that user will setup TLS for both upstream & downstream.

File: cdc/cdc/capture/http_validator.go [hunk 3]
```diff
@@ -132,6 +153,26 @@ func verifyUpdateChangefeedConfig(ctx context.Context, changefeedConfig model.Ch
...
+	if changefeedConfig.SortEngine != "" {
+		switch changefeedConfig.SortEngine {
+		case model.SortUnified, model.SortInMemory:
+		default:
+			return nil, cerror.ErrChangefeedUpdateRefused.GenWithStack("can not update sort engine: %s", changefeedConfig.SortEngine)
+		}
+		newInfo.Engine = changefeedConfig.SortEngine
+	}
+
+	if changefeedConfig.Format != "" || changefeedConfig.StartKey != "" || changefeedConfig.EndKey != "" {
```
[the remark this answers is not in the document]
Comment (zeminzhou) [reply cdc/cdc/capture/http_validator.go:165]: I modified as follow:
`if changefeedConfig.StartKey != "" || changefeedConfig.End != "" { 
    if changefeedConfig.Format == "" {
        changefeedConfig.Format = "hex"
    }
}` 
PTAL~

File: cdc/tests/.unmodify/http_api/util/test_case.py [hunk 4]
```diff
@@ -190,31 +188,31 @@ def remove_changefeed():
...
+    print("pass test: rebalance keyspan")
 
 
-def move_table():
-    # move table
-    url = BASE_URL0 + "/changefeeds/changefeed-test1/tables/move_table"
+def move_keyspan():
+    # move keyspan
+    url = BASE_URL0 + "/changefeeds/changefeed-test1/keyspans/move_keyspan"
     data = json.dumps({"capture_id": "test-aaa-aa", "table_id": 11})
```
[the remark this answers is not in the document]
Comment (zeminzhou) [reply cdc/tests/.unmodify/http_api/util/test_case.py:203]: modified, thanks!

File: cdc/tests/.unmodify/http_api/util/test_case.py [hunk 5]
```diff
@@ -190,31 +188,31 @@ def remove_changefeed():
...
-    print("pass test: rebalance table")
+    print("pass test: rebalance keyspan")
 
 
-def move_table():
-    # move table
-    url = BASE_URL0 + "/changefeeds/changefeed-test1/tables/move_table"
+def move_keyspan():
+    # move keyspan
+    url = BASE_URL0 + "/changefeeds/changefeed-test1/keyspans/move_keyspan"
```
[the remark this answers is not in the document]
Comment (zeminzhou) [reply cdc/tests/.unmodify/http_api/util/test_case.py:202]: Now we support move keyspan(verify manually), and I also add test to verfiy it. PTA~

File: cdc/tests/integration_tests/cli_tls/run.sh (hunk 1 above)
Comment (zeminzhou) [reply cdc/tests/integration_tests/cli_tls/run.sh:3]: modified, thanks!

File: cdc/tests/integration_tests/_utils/rawkv_op [hunk 6]
```diff
@@ -16,4 +16,4 @@ if ! command -v rawkv_data &>/dev/null; then
 fi
 
 echo "run put data"
-rawkv_data $2 --src-pd $1 --count $3 --ca-path=$TLS_DIR/ca.pem --cert-path=$TLS_DIR/client.pem --key-path=$TLD_DIR/client-key.pem
+rawkv_data $2 --src-pd $1 --count $3 --ca-path=$TLS_DIR/ca.pem --cert-path=$TLS_DIR/client.pem --key-path=$TLS_DIR/client-key.pem
```
[the remark this answers is not in the document]
Comment (zeminzhou) [reply cdc/tests/integration_tests/_utils/rawkv_op:19]: modified, thanks!

File: cdc/tests/integration_tests/cli_tls/run.sh (hunk 2 above)
Comment (zeminzhou) [reply cdc/tests/integration_tests/cli_tls/run.sh:12]: test on downstream with TLS has added [sink_tls](https://github.com/tikv/migration/tree/main/cdc/tests/integration_tests/sink_tls), PTAL~

Comment (pingyu) [reply cdc/tests/integration_tests/cli_tls/run.sh:12]: I mean setup TLS for both upstream & downstream at this case.
IMO users are more likely to enable TLS for both at the same time.

**Review by pingyu — COMMENTED** [review]
Rest LGTM.

File: cdc/cdc/capture/http_validator.go [hunk 7]
```diff
@@ -132,6 +153,29 @@ func verifyUpdateChangefeedConfig(ctx context.Context, changefeedConfig model.Ch
...
+	if changefeedConfig.SortEngine != "" {
+		switch changefeedConfig.SortEngine {
+		case model.SortUnified, model.SortInMemory:
+		default:
+			return nil, cerror.ErrChangefeedUpdateRefused.GenWithStack("can not update sort engine: %s", changefeedConfig.SortEngine)
+		}
+		newInfo.Engine = changefeedConfig.SortEngine
+	}
+
+	if changefeedConfig.StartKey != "" || changefeedConfig.EndKey != "" {
```
[the remark this answers is not in the document]
Comment (zeminzhou) [reply cdc/cdc/capture/http_validator.go:165]: Ok, add check to prohibit updating startKey & endKey

File: cdc/cdc/owner/scheduler_v1.go [hunk 8]
```diff
@@ -171,6 +171,7 @@ func (s *oldScheduler) handleMoveKeySpanJob() (shouldUpdateState bool, err error
 		// for all move keyspan job, here just remove the keyspan from the source capture.
 		// and the keyspan removed by this function will be added to target capture by syncKeySpansWithCurrentKeySpans in the next tick.
 		s.state.PatchTaskStatus(source, func(status *model.TaskStatus) (*model.TaskStatus, bool, error) {
+			log.Info("[ZZM] statue", zap.Any("status", status))
```
[the remark this answers is not in the document]
Comment (zeminzhou) [reply cdc/cdc/owner/scheduler_v1.go]: removed, thanks!

File: cdc/tests/integration_tests/cli_tls/run.sh (hunk 2 above)
Comment (zeminzhou) [reply cdc/tests/integration_tests/cli_tls/run.sh]: I also add  test for downstream with tls in `cli_tls` case

**Comment by zeminzhou** [comment]
@pingyu @haojinming PTAL~ again, I add check to prohibit updating startKey & endKey. Thanks！

File: cdc/pkg/cmd/cli/cli_changefeed_update.go [hunk 9]
```diff
@@ -192,12 +192,8 @@ func (o *updateChangefeedOptions) applyChanges(oldInfo *model.ChangeFeedInfo, cm
...
 			// Do nothing, this is a flags from the cli command
 			// we don't use it to update, but we do use these flags.
-		case flagStartKey:
-			newInfo.StartKey = o.commonChangefeedOptions.startKey
-		case flagEndKey:
-			newInfo.EndKey = o.commonChangefeedOptions.endKey
-		case flagKeyFormat:
-			newInfo.Format = o.commonChangefeedOptions.format
+		case flagStartKey, flagEndKey, flagKeyFormat:
+			log.Warn("this flag cannot be updated and will be ignored")
```
Comment (haojinming) [inline cdc/pkg/cmd/cli/cli_changefeed_update.go:196]: May need more detail info, e.g. flag name, flag value.

Comment (zeminzhou) [reply cdc/pkg/cmd/cli/cli_changefeed_update.go:196]: yesh, added, thanks!

File: cdc/tests/integration_tests/_utils/start_tls_tidb_cluster_impl [hunk 10]
```diff
@@ -45,26 +45,49 @@ cert-path = "$TLS_DIR/server.pem"
 key-path = "$TLS_DIR/server-key.pem"
 EOF
 
-echo "Starting TLS PD..."
+echo "Starting Upstream TLS PD..."
```
Comment (pingyu) [inline cdc/tests/integration_tests/_utils/start_tls_tidb_cluster_impl:48]: There are some duplicated codes between creating upstream & downstram cluster, and between creationg non-tls & tls clusters.

It's OK now, but suggest to consider extracting some core codes later. E.g, a function to start an single cluster, with arguments such as host/port of PD/TiDB/TiKV, `enable_tls`, and `log_prefix`.

**Pull request merged**